### DISCUSSION TITLE
how to create a new span based on a trace id?

### DISCUSSION THREAD

**Discussion by magicliang** [open-telemetry/opentelemetry-java#4668] [Q&A]
We have a producer-consumer architecture which uses an internal mq to pass messages.

We can add trace id (or, so-called parent trace id)to the message content or header in producer side, but how can we create a span from scratch based on that id, so a tracing context is propagated?

**Comment by mateuszrzeszutek** [comment]
Hey @magicliang ,
You could use the [context propagation API](https://opentelemetry.io/docs/instrumentation/java/manual/#context-propagation) to retrieve incoming trace context from your message headers -- this is exactly how other messaging instrumentation work (e.g. Kafka or JMS).
Alternatively, if the idea above does not suit your use case, you could manually construct a `SpanContext` from the raw traceId and spanId values using the [`createFromRemoteParent()`](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/trace/SpanContext.java#L78) method; and then wrap it in `Context.root().with(Span.wrap(<constructed SpanContext>))` to get a parent context for the consumer span.

**Reply by magicliang** [reply]
Hi mate,

I checked [https://opentelemetry.io/docs/instrumentation/java/manual/#context-propagation](https://opentelemetry.io/docs/instrumentation/java/manual/#context-propagation) before, I don't see how http api can be reuse in my context.

I also searched and found [createFromRemoteParent()](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/trace/SpanContext.java#L78) before, what I don't understand is what is correct format for parameter traceIdHex and spanIdHex?

An example I found is:

```java
   SpanContext parentContext =
          SpanContext.createFromRemoteParent(
              TraceId.fromLowerBase16(traceId, 0),
              SpanId.fromLowerBase16(spanId, 0),
              TraceFlags.builder().setIsSampled(false).build(),
              TraceState.builder().build());
```

can I use normal traceId in decimal base format as input for  TraceId.fromLowerBase16?

**Reply by mateuszrzeszutek** [reply]
> I checked ]https://opentelemetry.io/docs/instrumentation/java/manual/#context-propagation(https://opentelemetry.io/docs/instrumentation/java/manual/#context-propagation) before, I don't see how http api can be reuse in my context.

The HTTP API is completely not important in that example; the important parts are a `TextMapGetter` that can extract headers from your message, and extracting the context using `openTelemetry.getPropagators()`.

> I also searched and found [createFromRemoteParent()](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/trace/SpanContext.java#L78) before, what I don't understand is what is correct format for parameter traceIdHex and spanIdHex?

Please take a look at the `getTraceId()` and `getSpanId()` javadocs:

```java
  /**
   * Returns the trace identifier associated with this {@link SpanContext} as 32 character lowercase
   * hex String.
   *
   * @return the trace identifier associated with this {@link SpanContext} as lowercase hex.
   */
  String getTraceId();

  /**
   * Returns the span identifier associated with this {@link SpanContext} as 16 character lowercase
   * hex String.
   *
   * @return the span identifier associated with this {@link SpanContext} as 16 character lowercase
   *     hex (base16) String.
   */
  String getSpanId();
```

**Reply by magicliang** [reply]
> > I checked ]https://opentelemetry.io/docs/instrumentation/java/manual/#context-propagation(https://opentelemetry.io/docs/instrumentation/java/manual/#context-propagation) before, I don't see how http api can be reuse in my context.
> 
> The HTTP API is completely not important in that example; the important parts are a `TextMapGetter` that can extract headers from your message, and extracting the context using `openTelemetry.getPropagators()`.
> 
> > I also searched and found [createFromRemoteParent()](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/trace/SpanContext.java#L78) before, what I don't understand is what is correct format for parameter traceIdHex and spanIdHex?
> 
> Please take a look at the `getTraceId()` and `getSpanId()` javadocs:
> 
> ```java
>   /**
>    * Returns the trace identifier associated with this {@link SpanContext} as 32 character lowercase
>    * hex String.
>    *
>    * @return the trace identifier associated with this {@link SpanContext} as lowercase hex.
>    */
>   String getTraceId();
> 
>   /**
>    * Returns the span identifier associated with this {@link SpanContext} as 16 character lowercase
>    * hex String.
>    *
>    * @return the span identifier associated with this {@link SpanContext} as 16 character lowercase
>    *     hex (base16) String.
>    */
>   String getSpanId();
> ```

I believe Open Telemetry implementation can automatically inject context in http request and retrieve it from request ignoring what http client I am using. All programming work to be concerned is how to configure a custom TextMapGetter to openTelemetry.getPropagators() result(I guess there is a TextMapPropagator singleton). But I don't see how this automation can work while I am using a mq. I google for a long time, and see no auto-instrumentation implementation of Open Telemetry for Kafka or pulsar( only some relic of open tracing). I guess there be some support someday, but before that I have to use [createFromRemoteParent()] (https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/trace/SpanContext.java#L78) manually all by myself.

Or can you provide some hints about how to use TextMapGetter in a mq?
